### ISSUE TITLE
Add analyzer detecting & fixing use of Cast/OfType methods on generic IEnumerable sources.

### ISSUE THREAD

**Issue by kingmotley** [dotnet/runtime#77460]
### Description

The performance of the .OfType<T> and .Cast<T> methods are suboptimal in many cases.  For example, while essentially equivalent:

`people.Cast<Person>().ToList()` runs about 4 times as slow and takes 4 times the memory as `people.Select(x => (Person)x).ToList()` while being essentially logically equivalent.

// * Summary *

BenchmarkDotNet=v0.13.2, OS=Windows 11 (10.0.22000.856/21H2)
AMD Ryzen 9 5900X, 1 CPU, 24 logical and 12 physical cores
.NET SDK=6.0.203
  [Host]     : .NET 6.0.8 (6.0.822.36306), X64 RyuJIT AVX2
  DefaultJob : .NET 6.0.8 (6.0.822.36306), X64 RyuJIT AVX2

|        Method |      Mean |    Error |   StdDev | Allocated |
|-------------- |----------:|---------:|---------:|----------:|
|          Cast | 177.81 us | 1.554 us | 1.378 us | 256.41 KB |
|   Cast_Select |  43.62 us | 0.396 us | 0.331 us |  78.25 KB |
|        OfType | 193.63 us | 2.182 us | 2.041 us | 256.41 KB |
|     OfType_As |  97.13 us | 1.330 us | 1.244 us | 256.47 KB |
| OfExactlyType |  82.80 us | 0.805 us | 0.753 us | 256.47 KB |

### Configuration

Tested on .Net Core 6.0.8, windows 11 x64.

### Other information

CastIterator is on Enumerable, while the SelectListIterator/SelectArrayIterator/SelectEnumerableIterator is a generic iterator.  Creating optimized versions of the Cast/OfType methods could lead to significant performance increases.

[CastingObjects.zip](https://github.com/dotnet/core/files/9448367/CastingObjects.zip)

Credit for finding this should go to Nick Chapsas: https://youtu.be/dIu5EisoB_s

**Comment by eiriktsarpalis** [comment]
The difference in performance can be attribute to a couple of reasons:

1.  The `Select` method acts on generic `IEnumerable<T>` sources whereas `Cast` and `OfType` act on non-generic `IEnumerable` sources.
2. Your choice of source type is `List<T>`, which makes it possible for generic Linq methods to apply performance optimizations. If I change the input to a non-collection `IEnumerable<object>` then the difference effectively vanishes.

It is not possible for the `Cast` to apply any performance optimizations unless we expose an overload accepting `IEnumerable<T>` inputs, however such an overload would effectively negate the [purpose of the method](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.cast?view=net-7.0), which is to convert non-generic IEnumerable inputs to generic ones.

One possibility though is to add an analyzer that triggers whenever users call `Cast` on generic IEnumerable inputs (since it always performs worse than `Select`).

**Comment by kingmotley** [comment]
Changing the source to an IEnumerable<object> still shows performance improvements by using Cast_Select over the normal Cast for me:
|        Method |     Mean |     Error |    StdDev | Allocated |
|-------------- |---------:|----------:|----------:|----------:|
|          Cast | 2.528 ms | 0.0092 ms | 0.0072 ms |   1.55 MB |
|   Cast_Select | 1.887 ms | 0.0087 ms | 0.0081 ms |   1.37 MB |
|        OfType | 2.531 ms | 0.0059 ms | 0.0053 ms |   1.55 MB |
|     OfType_As | 2.488 ms | 0.0258 ms | 0.0241 ms |   1.55 MB |
| OfExactlyType | 2.488 ms | 0.0260 ms | 0.0243 ms |   1.55 MB |

**Comment by eiriktsarpalis** [comment]
Using select over range enumerables is still subject to performance optimizations. You should consider using an iterator method instead to ensure a (relatively) apples-to-apples comparison.

Like I said though, we won't be entertaining any changes to the Cast method itself, it is really only intended as an adapter for older non-generix APIs.

**Comment by kingmotley** [comment]
I think it is unfortunate that is your position considering that the .Cast method is already detecting that is an IEnumerable<T>:
```
        public static IEnumerable<
#nullable disable // there's no way to annotate the connection of the nullability of TResult to that of the source
                TResult
#nullable restore
                > Cast<TResult>(this IEnumerable source)
        {
            if (source is IEnumerable<TResult> typedSource)
            {
                return typedSource;
            }

            if (source == null)
            {
                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
            }

            return CastIterator<TResult>(source);
        }
```

**Comment by kingmotley** [comment]
Unfortunate considering how trivial of a change this would be, and would improve performance in many cases.

```
using System.Collections;
using BenchmarkDotNet.Attributes;

[MemoryDiagnoser(false)]
public class Benchmarks
{
    public record Person(Guid Id, string FullName);
    
    private static readonly IEnumerable<object> People = Enumerable
        .Range(0, 10000)
        .Select(x => (object)new Person(Guid.NewGuid(), Guid.NewGuid().ToString()));
  
    [Benchmark]
    public List<Person> SystemLinqCast()
    {
        return System.Linq.Enumerable.Cast<Person>(People)
            .ToList();
    }
    
    [Benchmark]
    public List<Person> BetterCast()
    {
        return People
            .Cast<Person>()
            .ToList();
    }
    
    [Benchmark]
    public List<Person> Cast_Select()
    {
        return People
            .Select(x => (Person)x)
            .ToList();
    }
}

public static class EnumerableExtensions
{
    public static IEnumerable<T> Cast<T>(this IEnumerable source)
    {
        if (source is IEnumerable<object> src)
        {
            return src.Select(x => (T)x);
        }

        return System.Linq.Enumerable.Cast<T>(source);
    }
}
```

Results:
|         Method |     Mean |     Error |    StdDev | Allocated |
|--------------- |---------:|----------:|----------:|----------:|
| SystemLinqCast | 2.530 ms | 0.0079 ms | 0.0062 ms |   1.55 MB |
|     BetterCast | 1.899 ms | 0.0114 ms | 0.0106 ms |   1.37 MB |
|    Cast_Select | 1.904 ms | 0.0161 ms | 0.0150 ms |   1.37 MB |

Essentially two lines of code if it were integrated into System.Linq.Enumerable.Cast directly.

**Comment by kingmotley** [comment]
That seems counter intuitive that the Cast method would not be the best choice for changing types on the `IEnumerable<T>` and the dotnet team feels that we should be using the less specific and less readable Select method, but if that is the opinion of the team then so be it.

The same goes for OfType.

I really hope the team reconsiders.  Perhaps adding documentation on these methods pointing out that while they work on `IEnumerable<T>`s, that they are specifically meant to only be for non-generics, and the negative performance implications?

**Comment by eiriktsarpalis** [comment]
> but if that is the opinion of the team then so be it

What optimizations would you suggest? As mentioned many of the optimizations in Select rely on the fact that the input source is generic which is not the case here.

**Comment by kingmotley** [comment]
Would it be possible to check if the source coming into the method is generic, and if so, then just calling the generic select to leverage it's already existing optimizations?

**Comment by eiriktsarpalis** [comment]
> Would it be possible to check if the source coming into the method is generic

Generic with respect to what element type? Cast only knows the type of the target enumerables

**Comment by eiriktsarpalis** [comment]
Correct, but I question whether such an overload would be useful beyond converting `IEnumerable<object>` sources, which aren't all that common. I think the best way forward is to consider an analyzer for improper uses of Cast.

**Comment by kingmotley** [comment]
Anecdotal, but going through a project that I am currently working on, in one instance we are using reflection to get Attributes on models (for validation).  The propertyInfo.GetCustomAttributes method that allows you to not inherit from parents returns an object array.  We then cast this object array into the Attribute type we just looked for, essentially `object[].Cast<CustomAttribute>()`.  Elsewhere, we are casting some Kendo objects from a derived type to a root parent type as we walk an expression tree and return all the leafs.

**Comment by eiriktsarpalis** [comment]
> The propertyInfo.GetCustomAttributes method that allows you to not inherit from parents returns an object array.

True, but I doubt Linq performance would be the bottleneck when running reflection code. Incidentally you might want to use one of the [generic extension methods](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.customattributeextensions.getcustomattributes?view=net-6.0#system-reflection-customattributeextensions-getcustomattributes-1(system-reflection-memberinfo-system-boolean)) to avoid needing to deal with object arrays.

> but if it is possible to just write the optimized code when dealing with a generic source, why would that not be preferable?

Generally speaking, Linq performance optimizations are a trade-off with System.Linq assembly size. We try to be strategic about what Linq methods we do optimize, and apply no optimizations at all in platforms where assembly size is important (e.g. wasm). You can see this in the implementation source files, some implementations are split between "speedopt" and "sizeopt" variants.

> In any case, thank you for your time Eirik.

You're welcome, happy to answer any other questions you may have.
